File: datamart/src/main/scala/DataMart.scala
import org.apache.spark.sql.{DataFrame, SparkSession}
import preprocess.Preprocessor
import db.MySqlDatabase
import logging.MyLoggerFactory
import config.ConfigLoader._

object DataMart {
  private val logger = MyLoggerFactory.getLogger(getClass)
  val session: SparkSession = SparkSession.builder
    .appName(spark.appName)
    .master(spark.deployMode)
    .config("spark.driver.cores", spark.driverCore)
    .config("spark.executor.cores", spark.executorCore)
    .config("spark.driver.memory", spark.driverMemory)
    .config("spark.executor.memory", spark.executorMemory)
    .config("spark.jars", s"${spark.mysqlConnector}:${spark.configS}")
    .config("spark.driver.extraClassPath", s"${spark.mysqlConnector}:${spark.configS}")
    .config("spark.driver.extraJavaOptions", spark.logger)
    .config("spark.executor.extraJavaOptions", spark.logger)
    .getOrCreate()

  private val db = new MySqlDatabase(session)

  def readPreprocessed(tablename: String): DataFrame = {
    logger.info("Reading and preprocessing OpenFoodFacts dataset")
    val data = db.readTable(tablename)
    val transforms: Seq[DataFrame => DataFrame] = Seq(
      Preprocessor.assembleVector,
      Preprocessor.scaleAssembledDataset
    )
    val transformed = transforms.foldLeft(data) { (df, f) => f(df) }
    logger.debug("Preprocessed dataset schema: " + transformed.schema)
    transformed
  }

  def writePredictions(df: DataFrame): Unit = {
    logger.info("Writing predictions to database")
    db.insertDf(df, "Predictions")
    logger.info("Predictions written to Predictions table")
  }
}

File: src/datamart.py
from pyspark.sql import DataFrame, SparkSession, SQLContext

from logger import Logger

SHOW_LOG = True


class DataMart:
    def __init__(self, spark: SparkSession):
        logger = Logger(SHOW_LOG)
        self.log = logger.get_logger(__name__)
        self.spark_context = spark.sparkContext
        self.sql_context = SQLContext(self.spark_context, spark)
        self.datamart = self.spark_context.jvm.DataMart
        self.log.info("Data Mart initialization")

    def read_dataset(self, tablename) -> DataFrame:
        jvm_data = self.datamart.readPreprocessed(tablename)
        self.log.info(f"Data Mart read dataset {tablename}")
        return DataFrame(jvm_data, self.sql_context)

    def write_predictions(self, df: DataFrame):
        self.log.info("Data Mart write predictions")
        self.datamart.writePredictions(df.jdf)